ocaml/parsing/jane_syntax.mli

```diff
@@ -165,13 +159,14 @@ module Mode_expr : sig
       attribute is found. *)
   val of_attrs : Parsetree.attributes -> t * Parsetree.attributes
 
-  (** Encodes a mode expression into a [payload]. If the expression is safe to
-      ignore (i.e. empty), returns [None]. *)
-  val payload_of : t -> Parsetree.payload option
+  (** Decode mode coercion and returns the mode and the body.
+      For example, return [Some (local, expr)] on input [local_ expr].
+      Returns [None] if the given expression is not a mode coercion. *)
+  val coerce_of_expr : Parsetree.expression -> (t * Parsetree.expression) option
```

Comment:
My impression was that we want to avoid using the full jane_syntax.

Reply:
I believe Zesen's motivation to make this PR was to make Alan's life easier in https://github.com/ocaml-flambda/flambda-backend/pull/2330. I think this PR accomplishes that.

Jane Syntax has some downsides. For example, using it would add another attribute to these expressions. That's not the main downside, though: the main downside is that there could be some amount of fighting Jane Syntax if we were to go the full-blown Jane Syntax route.

If we accept the extra attribute (that seems fine) and converting this to full-blown Jane Syntax is easy, then we can do that. But I do see this PR as an incremental improvement, so I'd prefer merging it if the Jane Syntax route proves hard.

Reply:
It's a judgement call: we want to avoid full jane_syntax for things that are just "extra bits of information". Full jane-syntax is designed for when you have to encode something wholly new (the first use-case was comprehensions). Adding modes for function arguments / returns is a case of "extra bits of information". That is, `'a @ local -> 'b` is still a function, not something wildly different. On the other hand, a mode coercion is really not an application at all. So it makes more sense to have the jane-syntax mechanism. It really shouldn't be hard to code up at all.

The rule of thumb I'm imagining here is this: when matching against a bit of AST, will you sometimes want to ignore the attribute? Or will you always want to look at it? If it's the latter then use full jane-syntax. I think that's the case here.

Curious for @ncik-roberts's opinion here, also.

Reply:
I'm not so worried about the extra attribute (which we could get rid of in general, but that's a fight for another day). I expect it would be easy to convert this to jane-syntax, and would simplify usage sites. I would suggest investing ~1 hour in trying it. If it fails, just merge this as is. If it succeeds, then use jane-syntax.

Just to add some motivation: I expect that using jane syntax here is somewhat likely to prevent some hard-to-find bug caused by someone forgetting to handle the mode-coercion case. (This is why jane syntax was invented!) Investing a little time now seems to have positive expected value.

Reply:
Thanks, Nick, for trying.